nautobot_device_lifecycle_mgmt/metrics.py

```diff
@@ -0,0 +1,57 @@
+"""Nautobot Device LCM  plugin application level metrics ."""
+from datetime import datetime
+
+from django.conf import settings
+from nautobot.dcim.models import Device, Site, InventoryItem
+from prometheus_client.core import GaugeMetricFamily
+
+from nautobot_device_lifecycle_mgmt.models import HardwareLCM
+
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG.get("nautobot_device_lifecycle_mgmt", {})
+
+
+def nautobot_metrics_dlcm_eos():
+    """Calculate number of EOS devices per Device Type and per Site.
+    Yields:
+        GaugeMetricFamily: Prometheus Metrics
+    """
+    current_dt = datetime.now()
+    hw_eos_notices = HardwareLCM.objects.filter(end_of_support__lte=current_dt)
+    hw_eos_device_types = [notice.device_type for notice in hw_eos_notices]
+    hw_eos_inventoryitems = [notice.inventory_item for notice in hw_eos_notices]
+
+    part_number_gauge = GaugeMetricFamily(
+        "nautobot_lcm_devices_eos_per_part_number", "Nautobot LCM Devices EOS per Part Number", labels=["part_number"]
+    )
+    devices_gauge = GaugeMetricFamily(
+        "nautobot_lcm_devices_eos_per_site", "Nautobot LCM Devices EOS per Site", labels=["site"]
+    )
+
+    for notice in hw_eos_notices:
+        if notice.device_type:
+            part_number = notice.device_type.part_number
+            eos_devices = Device.objects.filter(device_type=notice.device_type)
+            metric_value = eos_devices.count()
+        elif notice.inventory_item:
+            part_number = notice.inventory_item
+            eos_devices = InventoryItem.objects.filter(part_id=notice.inventory_item)
+            metric_value = eos_devices.count()
+        else:
+            continue
+
+        part_number_gauge.add_metric(labels=[part_number], value=metric_value)
+
+    yield part_number_gauge
```

Comment:
Since the `yield` is outside of the for loop, what will happen if there are no HW notices?

Reply:
Correct, the way I thought of these 2 metrics was to answer 2 typical net arch Questions when facing a refresh/netop initiative:
a) how many devices per site are out of vendor support. -> spot most compromised sites
b) what type of devices are end of support -> spot most compromised platforms

do you suggest adding a third metric that shows eos platforms unfiltered? Once a platform is EOS :
EOS devices for that platform == Platform.device.count()

Reply:
Actually no: I agree with the two questions that you're trying to answer.

That said (and if I'm not forgetting something), if a Site has _no_ EoS devices/inventory items, then it yields a `nautobot_lcm_devices_eos_per_site` metric line with a value of `0.0`. Something like:
```
nautobot_lcm_devices_eos_per_site{site="ang01"} 0.0
```

This is not the case with the `nautobot_lcm_devices_eos_per_part_number` metric: if there's no device with the specific part number, there's no yielded metric line. I would expect something like:
```
nautobot_lcm_devices_eos_per_part_number{part_number="DCS-7150S-24"} 0.0
```

Let me know if that's clear and if/how I could further assist you with this.